README.md
# yam2ncml.py
ROMS ncML aggregation from YAML especification


CHANGES.txt
0.6.1
~~~~~
- expose the CLI API

0.6.0
~~~~~
- Python 3 support.

0.5.0
~~~~~
- allow variables that depend on `Nbed` to be displayed by `ncWMS` by constructing a pseudo-depth coordinate.

0.4.0
~~~~~
- the list of `exclude` and `include` variables in the YAML now just control whether the 'display' attribute is 'true' or 'false'.  This means that all variables are still available on THREDDS.   Portals can use the 'display' attribute to show variables or not.

0.3.0
~~~~~
- the list of `exclude` and `include` variables in the YAML control whether the variables are retained, or removed via NcML.


tests/test_cli.py
from __future__ import (absolute_import, division, print_function)

import pytest
from docopt import DocoptExit
from docopt import docopt

from yaml2ncml import yaml2ncml

__doc__ = yaml2ncml.__doc__


def test_noarg_call():
    with pytest.raises(DocoptExit):
        yaml2ncml.main()


def test_mandatory_arg():
    fin = 'roms.yaml'
    args = docopt(__doc__, [fin])
    assert args['INFILE'] == fin


def test_optional_arg():
    fin = 'test6.ncml'
    fout = '--output=test6.ncml'
    args = docopt(__doc__, [fin, fout])
    assert args['--output'] == fout.split('=')[1]


tests/ROMS_yaml2ncml.py
# -*- coding: utf-8 -*-
# <nbformat>3.0</nbformat>

# <codecell>

import yaml
import pyncml
import netCDF4
import os
from StringIO import StringIO

# <codecell>

cd /usgs/data2/notebook/ROMS

# <codecell>

#map ROMS variables to CF standard_names
cf = {'zeta':'sea_surface_height_above_datum',
      'temp':'sea_water_potential_temperature',
      'salt':'sea_water_salinity',
         'u':'x_sea_water_velocity',
         'v':'y_sea_water_velocity',
      'ubar':'barotropic_x_sea_water_velocity',
      'vbar':'barotropic_y_sea_water_velocity',
     'Hwave':'sea_surface_wave_significant_height'}
        

# <codecell>

x="""

dataset:
    id: "USGS_COAWST_MVCO_CBLAST_Ripples_SWAN_40m"

    title: "USGS-CMG-COAWST Model: CBLAST2007 Ripples with SWAN-40m res"

    summary: "Simulation of hydrodynamics and bottom stress south of Marthas Vineyard, MA using the COAWST modeling system.  These results are from the 40m inner nest of a four-level nested simulation."
    
    creator:
        email: nganju@usgs.gov
        name: Neil Ganju
        url: http://water.usgs.gov/fluxes

      
    license: "The data may be used and redistributed for free but is not intended for legal use, since it may contain inaccuracies. Neither the data Contributor, nor the United States Government, nor any of their employees or contractors, makes any warranty, express or implied, including warranties of merchantability and fitness for a particular purpose, or assumes any legal liability for the accuracy, completeness, or usefulness, of this information."
    
    references: 
        - http://www.whoi.edu/science/AOPE/dept/CBLASTmain.html
        - http://water.usgs.gov/fluxes/mvco.html
        - doi:10.1029/2011JC007035

    acknowledgements:
        - USGS-CMGP
        - NSF    

variables: 
    include: 
        - temp
        - salt
 
    exclude: 
        - ubar
        - vbar

aggregation:
    time_var: ocean_time
    dir: /usgs/data0/mvco_ce/mvco_output/spatial_7_ar0fd
    sample_file: his_case7_ar0fd_0001.nc
    pattern: .*/his_case7_ar0fd_[0-9]{4}\.nc$


"""
#
# couldn't get this to work
#stream = open(StringIO(x))

# <codecell>

# so read file instead
stream = open("meta.yaml", 'r')
a = yaml.load(stream)

# <codecell>

a['dataset']

# <codecell>

def header():
    str='<?xml version="1.0" encoding="UTF-8"?>\n<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2">\n'
    str += str_att('Conventions','CF-1.6, SGRID-0.1, ACDD-1.3')
    str += str_att('cdm_data_type','Grid')
    return str

# <codecell>

def footer(str):
    str += '</netcdf>\n'
    return str

# <codecell>

def str_att(name,value):
    if isinstance(value, list):
        value = ','.join(value)
    return '  <attribute name="{:s}" type="String" value="{:s}"/>\n'.format(name,value)

# <codecell>

def add_global_atts(str,a):
    d = a['dataset']
    for key, value in d.iteritems():
        # handle simple attribute pairs first
        if key in ['id','license','summary','title','project','naming_authority','references','acknowledgements']:
            str += str_att(key,value)
            
        elif key in ['creator','publisher']:

            email = value.get("email", None)
            if email:
                str += str_att('_'.join([key,'email']),email)

            url = value.get("url", None)
            if url:
                str += str_att('_'.join([key,'url']),url)

            name = value.get("name", None)
            if name:
                str += str_att('_'.join([key,'name']),name)
                
        elif key in ['contributor']:
            role = value.get("role", None)
            if email:
                str += str_att('_'.join([key,'role']),role)
            
            email = value.get("email", None)
            if email:
                str += str_att('_'.join([key,'email']),email)

            url = value.get("url", None)
            if url:
                str += str_att('_'.join([key,'url']),url)

            name = value.get("name", None)
            if name:
                str += str_att('_'.join([key,'name']),name)
    return str

# <codecell>

def add_var_atts(str,a):
    ncfile=os.path.join(a['aggregation']['dir'],a['aggregation']['sample_file'])
    nc = netCDF4.Dataset(ncfile)
    ncv = nc.variables
    
    # get a list of all variables more than 1D
    vars = [var for var, vart in ncv.items() if vart.ndim > 1]
    
    vars_all = set(vars)
    vars_include = set(a['variables']['include'])
    vars_exclude = set(a['variables']['exclude'])
    if a['variables']['exclude']:
        vars = list(vars_all - vars_all.intersection(vars_exclude))
    else:
        if a['variables']['include']:
            vars = list(vars_all.intersection(vars_include))
        
    #rho_vars = [var for var, vart in ncv.items() if hasattr(vart,'coordinates') and 'lon_rho' in vart.coordinates and 'lat_rho' in vart.coordinates]    
    rho_vars = [var for var, vart in ncv.items() if 'eta_rho' in vart.dimensions and 'xi_rho' in vart.dimensions]
    u_vars = [var for var, vart in ncv.items() if 'eta_u' in vart.dimensions and 'xi_u' in vart.dimensions]
    v_vars = [var for var, vart in ncv.items() if 'eta_v' in vart.dimensions and 'xi_v' in vart.dimensions]
    
    for var in vars:
        str += '<variable name="{:s}">\n'.format(var)
        try:
            str += str_att('standard_name',cf[var])
        except:
            pass
        str += str_att('grid','grid')
        str += str_att('content_coverage_type','modelResult')
        if var in rho_vars:
            str += str_att('location','face')
        elif var in u_vars:
            str += str_att('location','edge1')
        elif var in v_vars:
            str += str_att('location','edge2')
        str += '</variable>\n\n'
        
    return str
    
    

# <codecell>

def write_grid_var(str):
    grid_var="""<variable name="grid" type="int">
        <attribute name="cf_role" value="grid_topology"/>
        <attribute name="topology_dimension" type="int" value="2"/>
        <attribute name="node_dimensions" value="xi_psi eta_psi"/>
        <attribute name="face_dimensions"
            value="xi_rho: xi_psi (padding: both) eta_rho: eta_psi (padding: both)"/>
        <attribute name="edge1_dimensions" value="xi_u: xi_psi eta_u: eta_psi (padding: both)"/>
        <attribute name="edge2_dimensions" value="xi_v: xi_psi (padding: both) eta_v: eta_psi"/>
        <attribute name="node_coordinates" value="lon_psi lat_psi"/>
        <attribute name="face_coordinates" value="lon_rho lat_rho"/>
        <attribute name="edge1_coordinates" value="lon_u lat_u"/>
        <attribute name="edge2_coordinates" value="lon_v lat_v"/>
        <attribute name="vertical_dimensions" value="s_rho: s_w (padding: none)"/>
    </variable>\n """
    str += grid_var
    return str

# <codecell>

def add_aggregation_scan(str,a):
    agg = a['aggregation']
#    <aggregation dimName="ocean_time" type="joinExisting">
#        <scan location="." regExp=".*his_case7_ar0fd_[0-9]{4}\.nc$" subdirs="false"/>
#    </aggregation>
    str += '<aggregation dimName="{:s}" type="joinExisting">\n'.format(agg['time_var'])
    str += '<scan location="." regExp="{:s}" subdirs="false"/>\n</aggregation>\n'.format(agg['pattern'])
    return str

# <codecell>

str = header()
str = add_global_atts(str,a)
str = add_var_atts(str,a)
str = write_grid_var(str)
str = add_aggregation_scan(str,a)
str = footer(str)

# <codecell>

print str

# <codecell>

with open('{:s}/test6.ncml'.format(a['aggregation']['dir']),'w') as text_file:
    text_file.write("{:s}".format(str))

# <codecell>




tests/test_yaml2ncml.py
from __future__ import (absolute_import, division, print_function)

import subprocess
import tempfile

import pytest
import ruamel.yaml as yaml

from yaml2ncml import build


def test_call():
    output = subprocess.check_output(['yaml2ncml', 'roms_0.yaml'])
    with open('base_roms_test.ncml') as f:
        expected = f.read()
    assert output.decode() == expected


def test_save_file():
    outfile = tempfile.mktemp(suffix='.ncml')
    subprocess.call(['yaml2ncml',
                     'roms_0.yaml',
                     '--output={}'.format(outfile)])
    with open('base_roms_test.ncml') as f:
        expected = f.read()
    with open(outfile) as f:
        output = f.read()
    assert output == expected


@pytest.fixture
def load_ymal(fname='roms_1.yaml'):
    with open(fname, 'r') as stream:
        yml = yaml.load(stream, Loader=yaml.RoundTripLoader)
    return yml


def test_bad_yaml():
    with pytest.raises(ValueError):
        yml = load_ymal(fname='roms_1.yaml')
        build(yml)


yaml2ncml/yaml2ncml.py
from __future__ import (absolute_import, division, print_function)

import os
import sys

import netCDF4
import ruamel.yaml as yaml

from six import raise_from
from docopt import docopt

__all__ = [
    'main',
    'build'
    ]

__doc__ = """
Generate ncml based on a yaml file.

Usage:
    yaml2ncml INFILE [--output=OUTFILE]

    yaml2ncml (-h | --help | --version)

Examples:
    yaml2ncml roms.yaml
    yaml2ncml roms.yaml --output=roms.ncml

Arguments:
  file      yaml file.

Options:
  -h --help     Show this screen.
  -v --version     Show version.
"""


def str_att(name, value):
    if isinstance(value, list):
        try:
            value = ','.join(value)
        except TypeError as e:
            raise_from(ValueError('Expected `str` got {!r}'.format(value)), e)
    msg = '  <attribute name="{:s}" type="String" value="{:s}"/>\n'
    return msg.format(name, value)


def header():
    text = '<?xml version="1.0" encoding="UTF-8"?>\n<netcdf xmlns='
    text += '"http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2">\n'
    text += str_att('Conventions', 'CF-1.6, SGRID-0.1, ACDD-1.3')
    text += str_att('cdm_data_type', 'Grid')
    return text


def footer(text):
    text += '</netcdf>\n'
    return text


def add_global_atts(text, a):
    d = a['dataset']
    for key, value in d.items():
        # Handle simple attribute pairs first.
        if key in ['id', 'license', 'summary', 'title', 'project',
                   'naming_authority', 'references', 'acknowledgments']:
            text += str_att(key, value)
        elif key in ['creator', 'publisher']:
            email = value.get("email", None)
            if email:
                text += str_att('_'.join([key, 'email']), email)
            url = value.get("url", None)
            if url:
                text += str_att('_'.join([key, 'url']), url)
            name = value.get("name", None)
            if name:
                text += str_att('_'.join([key, 'name']), name)
        elif key in ['contributor']:
            role = value.get("role", None)
            text += str_att('_'.join([key, 'role']), role)
            email = value.get("email", None)
            if email:
                text += str_att('_'.join([key, 'email']), email)
            url = value.get("url", None)
            if url:
                text += str_att('_'.join([key, 'url']), url)
            name = value.get("name", None)
            if name:
                text += str_att('_'.join([key, 'name']), name)
    return text


def add_bed_coord(text, a):
    ncfile = os.path.join(a['aggregation']['dir'],
                          a['aggregation']['sample_file'])
    nc = netCDF4.Dataset(ncfile)
    bed_coord_var = """<variable name="Nbed" shape="Nbed" type="double">
      <attribute name="long_name" value="pseudo coordinate at seabed points"/>
      <attribute name="standard_name" value="ocean_sigma_coordinate"/>
      <attribute name="positive" value="up"/>
      <attribute name="formula_terms" value="sigma: Nbed eta: zeta depth: h"/>
      <values start="-1.0" increment="-0.01"/>
    </variable>\n """
    if 'Nbed' in nc.dimensions.keys():
        text += bed_coord_var
    return text


def add_var_atts(text, a):
    ncfile = os.path.join(a['aggregation']['dir'],
                          a['aggregation']['sample_file'])
    nc = netCDF4.Dataset(ncfile)
    ncv = nc.variables

    # Get a list of all variables more than 1D.
    vars = [var for var, vart in ncv.items() if vart.ndim > 1]


#   identify all the rho, u and v vars

    rho_vars = [var for var in vars if 'eta_rho' in
                ncv[var].dimensions and 'xi_rho' in ncv[var].dimensions]
    u_vars = [var for var in vars if 'eta_u' in
              ncv[var].dimensions and 'xi_u' in ncv[var].dimensions]
    v_vars = [var for var in vars if 'eta_v' in
              ncv[var].dimensions and 'xi_v' in ncv[var].dimensions]

    vars_all = set(vars)
    vars_include = set(a['variables']['include'])
    vars_exclude = set(a['variables']['exclude'])

#   include/exclude only variables that actually occur in variable list

    vars_include = vars_all.intersection(vars_include)
    vars_exclude = vars_all.intersection(vars_exclude)

#   If there are variables excluded, exclude them and keep all rest.
#   If no variables are excluded, take just the included variables
#   If no variables are included or excluded, take all variables (leave
#     list of variables unchanged)

    if vars_exclude:
        vars_display = vars_all - vars_all.intersection(vars_exclude)
    else:
        if vars_include:
            vars_display = vars_all.intersection(vars_include)
        else:
            vars_display = vars_all

#   remove some variables we never want (if they exist)
    Tobc = set(['Tobc_in', 'Tobc_out'])
    vars_display = vars_display - vars_display.intersection(Tobc)

    vars_display = list(vars_display)

# add the variable attributes: S-grid stuff, display=T|F, ...
    for var in vars:
        text += '<variable name="{:s}">\n'.format(var)
        try:
            text += str_att('standard_name', cf[var])
        except:
            pass
        text += str_att('grid', 'grid')

        if 'Nbed' in ncv[var].dimensions:
            text += str_att('coordinates', ncv[var].coordinates+' Nbed')

        if var in vars_display:
            text += str_att('display', 'True')
        else:
            text += str_att('display', 'False')

        text += str_att('coverage_content_type', 'modelResult')
        if var in rho_vars:
            text += str_att('location', 'face')
        elif var in u_vars:
            text += str_att('location', 'edge1')
        elif var in v_vars:
            text += str_att('location', 'edge2')
        text += '</variable>\n\n'

# write standard_name for time coordinate variable
    var = 'ocean_time'
    if var in ncv.keys():
        try:
            text += '\n<variable name="{:s}">\n'.format(var)
            text += str_att('standard_name', cf[var])
            text += '</variable>\n\n'
        except:
            pass

    nc.close()
    return text


def write_grid_var(text):
    grid_var = """<variable name="grid" type="int">
        <attribute name="cf_role" value="grid_topology"/>
        <attribute name="topology_dimension" type="int" value="2"/>
        <attribute name="node_dimensions" value="xi_psi eta_psi"/>
        <attribute name="face_dimensions"
            value="xi_rho: xi_psi (padding: both) eta_rho: eta_psi (padding: both)"/>
        <attribute name="edge1_dimensions" value="xi_u: xi_psi eta_u: eta_psi (padding: both)"/>
        <attribute name="edge2_dimensions" value="xi_v: xi_psi (padding: both) eta_v: eta_psi"/>
        <attribute name="node_coordinates" value="lon_psi lat_psi"/>
        <attribute name="face_coordinates" value="lon_rho lat_rho"/>
        <attribute name="edge1_coordinates" value="lon_u lat_u"/>
        <attribute name="edge2_coordinates" value="lon_v lat_v"/>
        <attribute name="vertical_dimensions" value="s_rho: s_w (padding: none)"/>
    </variable>\n"""  # noqa
    text += grid_var
    return text


def add_aggregation_scan(text, a):
    agg = a['aggregation']
    text += '<aggregation dimName="{:s}" type="joinExisting">\n'.format(
        agg['time_var'])
    text += '<scan location="{:s}" regExp="{:s}" subdirs="false"/>\n</aggregation>\n'\
    .format(agg['dir'], agg['pattern'])  # noqa
    return text


# Map ROMS variables to CF standard_names. 
cf = dict(ocean_time='time',
          zeta='sea_surface_height_above_geopotential_datum',
          temp='sea_water_potential_temperature',
          salt='sea_water_salinity',
          u='sea_water_x_velocity',
          v='sea_water_y_velocity',
          ubar='barotropic_sea_water_x_velocity',
          vbar='barotropic_sea_water_y_velocity',
          Hwave='sea_surface_wave_significant_height',
          bed_thickness='sediment_bed_thickness')  #sediment_bed_thickness not in CF standard_names


def build(yml):
    text = header()
    text = add_global_atts(text, yml)
    text = add_var_atts(text, yml)
    text = write_grid_var(text)
    text = add_bed_coord(text, yml)
    text = add_aggregation_scan(text, yml)
    text = footer(text)
    return text


def main():
    args = docopt(__doc__, version='0.6.0')
    fname = args.get('INFILE')
    fout = args.get('--output', None)

    with open(fname, 'r') as stream:
        yml = yaml.load(stream, Loader=yaml.RoundTripLoader)

    text = build(yml)

    if fout:
        with open(fout, 'w') as f:
            f.write("{:s}".format(text))
    else:
        sys.stdout.write(text)


yam2ncml.py
import os
import yaml
import pyncml
import netCDF4
from StringIO import StringIO



#map ROMS variables to CF standard_names
cf = {'zeta':'sea_surface_height_above_datum',
      'temp':'sea_water_potential_temperature',
      'salt':'sea_water_salinity',
         'u':'x_sea_water_velocity',
         'v':'y_sea_water_velocity',
      'ubar':'barotropic_x_sea_water_velocity',
      'vbar':'barotropic_y_sea_water_velocity',
     'Hwave':'sea_surface_wave_significant_height'}
        



x = """
dataset:
    id: "USGS_COAWST_MVCO_CBLAST_Ripples_SWAN_40m"

    title: "USGS-CMG-COAWST Model: CBLAST2007 Ripples with SWAN-40m res"

    summary: "Simulation of hydrodynamics and bottom stress south of Marthas Vineyard, MA using the COAWST modeling system.  These results are from the 40m inner nest of a four-level nested simulation."
    
    creator:
        email: nganju@usgs.gov
        name: Neil Ganju
        url: http://water.usgs.gov/fluxes

    publisher:
        email: tkalra@usgs.gov
        name: Tarandeep Kalra
        url: http://www.usgs.gov
        
    contributor:
        role: advisor
        email: rsignell@usgs.gov
        name: Rich Signell
        url: http://profile.usgs.gov/rsignell

        
    license: "The data may be used and redistributed for free but is not intended for legal use, since it may contain inaccuracies. Neither the data Contributor, nor the United States Government, nor any of their employees or contractors, makes any warranty, express or implied, including warranties of merchantability and fitness for a particular purpose, or assumes any legal liability for the accuracy, completeness, or usefulness, of this information."
    
    references: 
        - http://www.whoi.edu/science/AOPE/dept/CBLASTmain.html
        - http://water.usgs.gov/fluxes/mvco.html
        - doi:10.1029/2011JC007035

    acknowledgements:
        - USGS-CMGP
        - NSF    

variables: 
    include: 
        - temp
        - salt
 
    exclude: 
        - ubar
        - vbar

aggregation:
    time_var: ocean_time
    dir: /usgs/data0/mvco_ce/mvco_output/spatial_7_ar0fd
    sample_file: his_case7_ar0fd_0001.nc
    pattern: .*/his_case7_ar0fd_[0-9]{4}\.nc$
"""

# Couldn't get this to work
# stream = open(StringIO(x))

# so read file instead
with open("meta.yaml", 'r') as stream:
    a = yaml.load(stream)

a['dataset']

def header():
    str='<?xml version="1.0" encoding="UTF-8"?>\n<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2">\n'
    str += str_att('Conventions','CF-1.6, SGRID-0.1, ACDD-1.3')
    str += str_att('cdm_data_type','Grid')
    return str

def footer(str):
    str += '</netcdf>\n'
    return str

def str_att(name,value):
    if isinstance(value, list):
        value = ','.join(value)
    return '  <attribute name="{:s}" type="String" value="{:s}"/>\n'.format(name,value)


def add_global_atts(str,a):
    d = a['dataset']
    for key, value in d.iteritems():
        # handle simple attribute pairs first
        if key in ['id','license','summary','title','project','naming_authority','references','acknowledgements']:
            str += str_att(key,value)
            
        elif key in ['creator','publisher']:

            email = value.get("email", None)
            if email:
                str += str_att('_'.join([key,'email']),email)

            url = value.get("url", None)
            if url:
                str += str_att('_'.join([key,'url']),url)

            name = value.get("name", None)
            if name:
                str += str_att('_'.join([key,'name']),name)
                
        elif key in ['contributor']:
            role = value.get("role", None)
            if email:
                str += str_att('_'.join([key,'role']),role)
            
            email = value.get("email", None)
            if email:
                str += str_att('_'.join([key,'email']),email)

            url = value.get("url", None)
            if url:
                str += str_att('_'.join([key,'url']),url)

            name = value.get("name", None)
            if name:
                str += str_att('_'.join([key,'name']),name)
    return str

def add_var_atts(str,a):
    ncfile=os.path.join(a['aggregation']['dir'],a['aggregation']['sample_file'])
    nc = netCDF4.Dataset(ncfile)
    ncv = nc.variables
    
    # get a list of all variables more than 1D
    vars = [var for var, vart in ncv.items() if vart.ndim > 1]
    
    vars_all = set(vars)
    vars_include = set(a['variables']['include'])
    vars_exclude = set(a['variables']['exclude'])
    if a['variables']['exclude']:
        vars = list(vars_all - vars_all.intersection(vars_exclude))
    else:
        if a['variables']['include']:
            vars = list(vars_all.intersection(vars_include))
        
    #rho_vars = [var for var, vart in ncv.items() if hasattr(vart,'coordinates') and 'lon_rho' in vart.coordinates and 'lat_rho' in vart.coordinates]    
    rho_vars = [var for var, vart in ncv.items() if 'eta_rho' in vart.dimensions and 'xi_rho' in vart.dimensions]
    u_vars = [var for var, vart in ncv.items() if 'eta_u' in vart.dimensions and 'xi_u' in vart.dimensions]
    v_vars = [var for var, vart in ncv.items() if 'eta_v' in vart.dimensions and 'xi_v' in vart.dimensions]
    
    for var in vars:
        str += '<variable name="{:s}">\n'.format(var)
        try:
            str += str_att('standard_name',cf[var])
        except:
            pass
        str += str_att('grid','grid')
        str += str_att('content_coverage_type','modelResult')
        if var in rho_vars:
            str += str_att('location','face')
        elif var in u_vars:
            str += str_att('location','edge1')
        elif var in v_vars:
            str += str_att('location','edge2')
        str += '</variable>\n\n'
        
    return str

def write_grid_var(str):
    grid_var="""<variable name="grid" type="int">
        <attribute name="cf_role" value="grid_topology"/>
        <attribute name="topology_dimension" type="int" value="2"/>
        <attribute name="node_dimensions" value="xi_psi eta_psi"/>
        <attribute name="face_dimensions"
            value="xi_rho: xi_psi (padding: both) eta_rho: eta_psi (padding: both)"/>
        <attribute name="edge1_dimensions" value="xi_u: xi_psi eta_u: eta_psi (padding: both)"/>
        <attribute name="edge2_dimensions" value="xi_v: xi_psi (padding: both) eta_v: eta_psi"/>
        <attribute name="node_coordinates" value="lon_psi lat_psi"/>
        <attribute name="face_coordinates" value="lon_rho lat_rho"/>
        <attribute name="edge1_coordinates" value="lon_u lat_u"/>
        <attribute name="edge2_coordinates" value="lon_v lat_v"/>
        <attribute name="vertical_dimensions" value="s_rho: s_w (padding: none)"/>
    </variable>\n """
    str += grid_var
    return str

def add_aggregation_scan(str,a):
    agg = a['aggregation']
#    <aggregation dimName="ocean_time" type="joinExisting">
#        <scan location="." regExp=".*his_case7_ar0fd_[0-9]{4}\.nc$" subdirs="false"/>
#    </aggregation>
    str += '<aggregation dimName="{:s}" type="joinExisting">\n'.format(agg['time_var'])
    str += '<scan location="." regExp="{:s}" subdirs="false"/>\n</aggregation>\n'.format(agg['pattern'])
    return str

str = header()
str = add_global_atts(str,a)
str = add_var_atts(str,a)
str = write_grid_var(str)
str = add_aggregation_scan(str,a)
str = footer(str)

print(str)

with open('{:s}/test4.ncml'.format(a['aggregation']['dir']),'w') as text_file:
    text_file.write("{:s}".format(str))
